### PR TITLE
add support for multiplex input validation against a requirements file

### PR DESCRIPTION
**Pull request by k-rister**
- move the check that validates the benchmark subproject earlier --
  this ensures it is there when we need it and also avoids doing
  unecessary work (if it is going to fail) before performing this
  validation step

- detect if the benchmark subproject has a multiplex requirements file
  and use it if it does

- rework the logging and execution a bit to share the same command
  variable so the user is not forced to synchronize the two

### CODE REVIEW

**Comment by k-rister** [comment]
I don't think we can merge this until we have a bit more confidence in the new multiplex requirements validation, but at the same time this code is needed to actually exercise this new capability so I'm posting it in draft status for the time being.

The code detects if a benchmark subproject has a requirements file and uses it in the multiplex invocation if it does.  For most benchmark subprojects there is no requirements file, however the fio and uperf subprojects have one that is not compatible with the new format so if this code where to be merged right now it would break fio and uperf.  I have a working draft of the fio requirements file but the port is incomplete due to multiple issues in multiplex (for example, I cannot port the value transformation piece because multiplex does not support that yet).  I believe there are multiplex issues covering all the things that are missing at this point.

File: bin/_main [hunk 1]
```diff
@@ -334,18 +334,35 @@ elif [ "${1}" == "run" ]; then
...
             mv_params_run_dir=${base_run_dir}/config/mv-params.json
             cp ${mv_params} ${mv_params_run_dir}
             bench_params_run_dir=${base_run_dir}/config/bench-params.json
 
-            echo "Generating --bench-params from --mv-params..."
-            echo "$podman_run -i --name crucible-multiplex "${container_common_args[@]}" $CRUCIBLE_CONTAINER_IMAGE ${CRUCIBLE_HOME}/subprojects/core/multiplex/multiplex.py --input ${mv_params_run_dir} > ${bench_params_run_dir}"
-            $podman_run -i --name crucible-multiplex "${container_common_args[@]}" $CRUCIBLE_CONTAINER_IMAGE ${CRUCIBLE_HOME}/subprojects/core/multiplex/multiplex.py --input ${mv_params_run_dir} > ${bench_params_run_dir}
+            multiplex_cmd="${CRUCIBLE_HOME}/subprojects/core/multiplex/multiplex.py --input ${mv_params_run_dir}"
+            if [ -e ${benchmark_subproj_dir}/multiplex.json ]; then
+                multiplex_cmd+=" --requirements ${benchmark_subproj_dir}/multiplex.json"
```
Comment (rafaelfolco) [inline bin/_main:359]: requirements file should be manually copied by the user to this dir ?

**Review by rafaelfolco — DISMISSED** [review]
lgtm, open questions on how it should be (we can improve docs)

File: bin/_main (hunk 1 above)
Comment (k-rister) [reply bin/_main:359]: no, the requirements file is part of the benchmark subproject -- ie. bench-fio has its own, bench-uperf has its own, etc.

File: bin/_main [hunk 2]
```diff
@@ -334,18 +334,35 @@ elif [ "${1}" == "run" ]; then
...
+            else
+                echo "Skipping benchmark requirements validation because there is no ${benchmark_subproj_dir}/multiplex.json"
+            fi
+
+            echo "$podman_run -i --name crucible-multiplex "${container_common_args[@]}" $CRUCIBLE_CONTAINER_IMAGE ${multiplex_cmd} > ${bench_params_run_dir}"
+            $podman_run -i --name crucible-multiplex "${container_common_args[@]}" $CRUCIBLE_CONTAINER_IMAGE ${multiplex_cmd} > ${bench_params_run_dir}
```
[the remark this answers is not in the document]
Comment (k-rister) [reply bin/_main:365]: STDERR is going to the console and informs the user that the attempted run is broken, there is no need to redirect to a file

[the remark this answers is not in the document]
Comment (k-rister) [reply bin/_main]: I'm working on doing something with this but multiplex bugs blocked me for the moment...

[the remark this answers is not in the document]
Comment (k-rister) [reply bin/_main]: I have a new approach for this.  I switch to using `--output` and then captured all STDOUT/STDERR together and dump that if a non-zero return code is observed.

**Comment by k-rister** [comment]
We should probably review and merge https://github.com/perftool-incubator/bench-uperf/pull/30 before this one.

**Review by rafaelfolco — APPROVED** [review]
lgtm

**Pull request merged**